jofing-forDevReview/ComponentDialogs/Dialogs/CheckInDialog.cs

```diff
@@ -0,0 +1,92 @@
+﻿namespace ComponentDialogs
+{
+    using Microsoft.Bot.Builder;
+    using Microsoft.Bot.Builder.Dialogs;
+    using Microsoft.Recognizers.Text;
+
+    /// <summary>
+    /// Defines the guest check-in dialog.
+    /// </summary>
+    public class CheckInDialog : ComponentDialog
+    {
+        /// <summary>
+        /// State information associated with this dialog.
+        /// </summary>
+        public class GuestInfo
+        {
+            public string Name { get; set; }
+            public int Room { get; set; }
+        }
+
+        /// <summary>
+        /// IDs for the prompts used in this dialog.
+        /// </summary>
+        private struct PromptIds
+        {
+            public const string GuestName = "textPrompt";
+            public const string RoomNumber = "numberPrompt";
+        }
+
+        /// <summary>
+        /// IDs for the values stored in dialog state.
+        /// </summary>
+        private class Values
+        {
+            public const string GuestInfo = "guestInfo";
+        }
+
+        /// <summary>
+        /// Creates a new instance of the <see cref="CheckInDialog"/> class.
+        /// </summary>
+        /// <param name="dialogId">The ID to assign to the main dialog in the set.</param>
+        public CheckInDialog(string dialogId) : base(dialogId)
+        {
+            // Indicate which dialog is the main dialog.
+            InitialDialogId = Id;
+
+            // Define the prompts used in this conversation flow.
+            AddDialog(new TextPrompt(PromptIds.GuestName));
+            AddDialog(new NumberPrompt<int>(PromptIds.RoomNumber, defaultLocale: Culture.English));
+
+            // Define the conversation flow using a waterfall model.
+            AddDialog(new WaterfallDialog(Id, new WaterfallStep[]
+            {
+                async (dc, step, cancellationToken) =>
+                {
+                    // Initialize guest information and prompt for the guest's name.
+                    step.Values[Values.GuestInfo] = new GuestInfo();
+                    return await dc.PromptAsync(PromptIds.GuestName, new PromptOptions
+                    {
+                        Prompt = MessageFactory.Text("What is your name?")
+                    });
```

Comment:
we should not use inline lamdas in waterfalls, it is clearer if these are pulled out into methods

Reply:
Nearly all of our current dialog samples use lambdas for waterfall steps, partly because it is less wordy and easier to read as rendered on the MS docs site. I'll bring this up in our stand up on Monday.

Reply:
In the multi-turn bot which is a demonstration of how to construct a waterfall I pulled the logic out into separate functions. However, in the Custom-Dialog samples I just wanted a two step waterfall to drive the scenario.

In other words I think both are valid but if its a non-trivial waterfall then I think I'd agree with Tom and pull it into functions. I think, however, its actually good to show people that both techniques are valid and both can be good C#.

One interesting observation is that I find it quite hard to name the functions in a waterfall. The problem is the function itself doesn't really align with a concept in the domain. Its not exactly an event handler, nor is it a processing step as you might imagine it in a flow chart. In a step function you are processing the results of a previous step and starting anew step. We've seen people get confused between prompt validation functions and waterfall steps functions. I think the confusing is reasonable; this is an odd way to write code.